Python/bytecodes.c

```diff
@@ -3445,31 +3759,35 @@ dummy_func(
 
         inst(EXIT_INIT_CHECK, (should_be_none -- )) {
             assert(STACK_LEVEL() == 2);
-            if (should_be_none != Py_None) {
+            PyObject *should_be_none_o = PyStackRef_AsPyObjectBorrow(should_be_none);
+            if (should_be_none_o != Py_None) {
                 PyErr_Format(PyExc_TypeError,
                     "__init__() should return None, not '%.200s'",
-                    Py_TYPE(should_be_none)->tp_name);
+                    Py_TYPE(should_be_none_o)->tp_name);
                 ERROR_NO_POP();
             }
         }
 
         op(_CALL_BUILTIN_CLASS, (callable, self_or_null, args[oparg] -- res)) {
+            PyObject *callable_o = PyStackRef_AsPyObjectBorrow(callable);
+
             int total_args = oparg;
-            if (self_or_null != NULL) {
+            if (!PyStackRef_IsNull(self_or_null)) {
                 args--;
                 total_args++;
             }
-            DEOPT_IF(!PyType_Check(callable));
-            PyTypeObject *tp = (PyTypeObject *)callable;
+            DEOPT_IF(!PyType_Check(callable_o));
+            PyTypeObject *tp = (PyTypeObject *)callable_o;
             DEOPT_IF(tp->tp_vectorcall == NULL);
             STAT_INC(CALL, hit);
-            res = tp->tp_vectorcall((PyObject *)tp, args, total_args, NULL);
+            PyObject *res_o = PyObject_TypeVectorcall_StackRef(tp, args, total_args, NULL);
```

Comment:
Discussed online elsewhere -- not possible, needs to untag out of place.

Reply:
Why are you saying this is not possible, when I have literally written the code for?

If it helps, here are outline free threading implementations of those functions:
```C
PyObject **
PyObjectArray_FromStackRefArray(_PyStackRef *input, int size, PyObject **scratch)
{
    PyObject **result;
    if (size > MAX_UNTAG_SCRATCH) {
        result = malloc(...);
    }
    else {
        result = scratch;
    }
    for (int i = 0; i < size; i++) {
        result[i] = UNTAG(input[i]);
    }
    return result;
}

void
PyObjectArray_Free(PyObject **array, PyObject **scratch)
{
    if (array != scratch) free(array);
}
```

Reply:
Woops sorry I misread your original post. Yes this is possible, in fact, I think it will work even without the usual idef gil disabled stuff.

The only part that might block this is that we would consume even more C stack space on the default build for Windows and WASI. The current way of abstracting it away completely removes any possible overhead ` PyObject *temp[MAX_UNTAG_SCRATCH];` might take up. The compiler cant DCE this away either because of `PyObjectArray_Free` using the scratch space.

Reply:
Nonetheless, I will try it out and see if it works.